feat: automatic completion of internal model migration

kian99 opened the pull request:
## Description
This PR aims to make the completion of an internal model migration automatic, it builds on #1669. 

Currently a model migration between two controller managed by JIMM must be completed by the user using the `update-migrated-model` command of the `jaas` plugin. With this change we adjust the method `CleanupNotFoundModels` into `PollModels` since this method was being run periodically to identify when models were deleted and remove them from JIMM's database.

The new `PollModels` method does the same thing as before (requests ModelInfo for all models) with a slight optimisation in that we dial each controller only once instead of dialing the same controller repeatedly (not an issue before since we had the dial cache). Next, the logic is based on the model's migration mode, instead of skipping the model if it was in a migrating state, now we can complete the migration if its state is MigrationInternal.

Rather than creating a separate method to poll for internally migrating models I've chosen to put the logic together with polling for deleted models because of the following reasons:
- `CleanupNotFoundModels` was already called every 1 minute which is adequate for completing internal migrations.
- The polling is doing the same (calling `ModelInfo`).

If we want to split these up so that one can run more frequently than the other we can do that in the future.

Finally, ideally we'd have a real integration test in our bash suite for internal model migration, I'll add this in a follow-up.

Fixes [JUJU-7481](https://warthogs.atlassian.net/browse/JUJU-7481)

## Engineering checklist

- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests

## Test instructions
Successfully tested this locally,
1. Start JIMM `docker compose up -d --wait`
2. Add controller to JIMM `./local/jimm/setup-controller.sh` and `./local/jimm/add-controller.sh`
3. Add another controller with a different name, `export CONTROLLER_NAME=test-controller` and then repeat step 2.
5. `juju add-model test-1` and `juju show-model` and record the model UUID.
6. Switch to the controllers `qa-lxd` and `test-controller` to see where it landed.
8. `go build ./cmd/jaas`
9. Depending on where the model landed `./jaas migrate-internal <controller> <model-UUID>`
10. Run `juju status` after the migration starts and you should get an `unauthorized` error. Wait 1 minute and then `juju status` should succeed.
11. (optionally) Watch logs with `docker logs -f jimm`.


[JUJU-7481]: https://warthogs.atlassian.net/browse/JUJU-7481?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kian99 committed:
feat: automatic completion of internal model migration

kian99 committed:
refactor: use controller alias over controller tag

When Juju returns the redirect error, it doesn't return the controller tag that was provided but it does return the controller alias.
Because controller names are unique in JIMM, we can provide Juju with the controller alias and use it once the migration succeeds.

alesstimec commented on internal/db/model_test.go:
```diff
@@ -766,3 +766,77 @@ func (s *dbSuite) TestCountModelsByController(c *qt.C) {
+  controller: test
+`
+
+func (s *dbSuite) TestSetModelMigrating(c *qt.C) {
```
TestSetModelMigrationMode

alesstimec commented on internal/db/model_test.go:
```diff
@@ -766,3 +766,77 @@ func (s *dbSuite) TestCountModelsByController(c *qt.C) {
+	c.Assert(migrationMode, qt.Equals, dbmodel.MigrationModeMigrateInternal)
+}
+
+func (s *dbSuite) TestSetModelMigrating_ModelDoesNotExist(c *qt.C) {
```
SetModelMigrationMode_ModelDoesNotExist

alesstimec commented on internal/db/model_test.go:
```diff
@@ -766,3 +766,77 @@ func (s *dbSuite) TestCountModelsByController(c *qt.C) {
+	c.Assert(err, qt.ErrorMatches, `model with uuid "fake-uuid" does not exist`)
+}
+
+func (s *dbSuite) TestSetModelMigrating_PreventSetAgain(c *qt.C) {
```
SetModelMigrationMode_PreventSetAgain

alesstimec commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,145 @@
+		}
+		api, err := j.dialController(ctx, &models[0].Controller)
+		if err != nil {
+			zapctx.Error(ctx, "cannot dial controller", zap.String("controller", controllerUUID), zap.Error(err))
```
if somebody bypassed JIMM and killed the controller, we would never clean up those models, right?

alesstimec replied:
how would we handle this situation? we might have to provide some tooling to resolve it..

kian99 replied:
Yip that's true. Nothing for this PR imo though.

alesstimec reviewed (commented):
LGTM in general, but i'm not too fond of the tight coupling between two orthogonal tasks: model cleanup and finishing model migrations, based on the migration mode of the model..

kian99 committed:
feat: replace SetModelMigrationMode with GetModelForUpdate

The SetModelMigrationMode method was placing logic in the DB layer and after some comments on the first PR I realised that i was introducing inconsistent behavior where some DB logic was captured within the method and other logic was captured on the dbmodel struct.
This change introduces a GetModelForUpdate method that will return the model and lock the row for updates, allowing the caller to handle the business logic while adding some helper methods on the dbmodel.Model struct.

ale8k commented on internal/db/model_test.go:
```diff
@@ -716,6 +718,44 @@ func (s *dbSuite) TestGetModelsByController(c *qt.C) {
 }
 
+func (s *dbSuite) TestGetModelForUpdate(c *qt.C) {
+	ctx := context.Background()
```
c.Context() here no?

ale8k commented on internal/db/model.go:
```diff
@@ -48,6 +46,37 @@ func (d *Database) GetModel(ctx context.Context, model *dbmodel.Model) (err erro
+
+// GetModelForUpdate retrieves a model and locks the row for updates.
+// This method should be used within a transaction.
+func (d *Database) GetModelForUpdate(ctx context.Context, model *dbmodel.Model) (err error) {
```
Can we append Tx to the name? I did that for some that should be used in transactions just to differentiate?

kian99 replied:
Yes for sure.

ale8k commented on internal/db/model_test.go:
```diff
@@ -716,6 +718,44 @@ func (s *dbSuite) TestGetModelsByController(c *qt.C) {
+			c.Check(err, qt.IsNil)
+
+			close(rowLocked)
+			<-ctx.Done()
```
Just use a context here again for consistency with the rowLocked chan, looks kinda strange using ctx to co-ordinate channels here and above using rowLocked

kian99 replied:
I'll use two channel rather then.

ale8k commented on internal/db/model.go:
```diff
@@ -48,6 +46,37 @@ func (d *Database) GetModel(ctx context.Context, model *dbmodel.Model) (err erro
+// GetModelForUpdate retrieves a model and locks the row for updates.
+// This method should be used within a transaction.
+func (d *Database) GetModelForUpdate(ctx context.Context, model *dbmodel.Model) (err error) {
+	const op = errors.Op("db.GetModelForUpdate")
```
Looking over this a few times... I'm not keen on having a GetModelForUpdate. I would've expected something more like...

```go
func (d *Database) ForUpdate(ctx context.Context) *Database {
	if forUpdate {
		db = db.Clauses(clause.Locking{Strength: "UPDATE"})
	}
}

// And somewhere in the code (pseudo)...
// db.ForUpdate().GetModel().UpdateModel()
```

kian99 replied:
That looks cool but seems like a very different paradigm to our current setup.

kian99 replied:
Added it. I kept the test and it continues to work. I also verified that the additional queries as a result of Preload() don't have a `FOR UPDATE` clause and they don't seem to. Fwiw the SQL queries generated from the test `TestGetModelForUpdate` can be seen here https://pastebin.ubuntu.com/p/Bc2qQ339pZ/

ale8k commented on internal/dbmodel/model.go:
```diff
@@ -182,3 +182,29 @@ func (m Model) MergeModelSummaryFromController(modelSummaryFromController *jujup
+
+// ProcessFailedMigration processes a failed migration by resetting the
+// MigrationMode to MigrationModeNone.
+func (m *Model) ProcessFailedMigration() {
```
This is beautiful

ale8k commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,145 @@
+// PollModels loops over models, contacting the respective controller
+// and checking, based on the model's migration mode, if the model exists.
+func (j *JujuManager) PollModels(ctx context.Context) (err error) {
+	const op = errors.Op("jimm.CleanupNotFoundModels")
```
OpCode

ale8k commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,145 @@
+
+	// Step 1: Group models by controller
+	controllerModels := make(map[string][]*dbmodel.Model)
+	err = j.Database.ForEachModel(ctx, func(m *dbmodel.Model) error {
```
Could we preload models on the controller here instead of looping over?

kian99 replied:
Do you mean you want like `SELECT * CONTROLLERS` instead of this which is `SELECT * MODELS`?

ale8k replied:
More like
```go
	d.ForEachController(ctx, func(ccontroller *dbmodel.Controller) error {
		// Models not preloaded though
		controllerModels[ccontroller.Name] = ccontroller.Models
		return nil
	})
```

So additionally something like:
```go
 ForEachController(ctx context.Context, f func(*dbmodel.Controller) error, preloadModels bool)
```

And I think it'd read nicer as it just wants the models, for each controller.

kian99 replied:
I think it's pretty equivalent so I'll leave this one as is.

ale8k commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,145 @@
+)
+
+// PollModels loops over models, contacting the respective controller
+// and checking, based on the model's migration mode, if the model exists.
```
based on the model's migration mode, if the model exists. 
If the model exists in JIMM's database, but not on the controller,
it is deleted from JIMM's database.

ale8k commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,145 @@
+// This performs eventual cleanup of models that have been deleted through
+// the API (since the deletion of a model is not immediate) and handles
+// cases where the model was deleted directly on the underlying controller.
+func (j *JujuManager) checkModelExists(ctx context.Context, api API, m *dbmodel.Model) {
```
maybeCleanupModel would be a better name

ale8k commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,145 @@
+// checkModelMigratedInternal checks if the model has been migrated from
+// one controller managed by JIMM to another controller managed by JIMM.
+func (j *JujuManager) checkModelMigratedInternal(ctx context.Context, api API, m *dbmodel.Model) {
+	const op = errors.Op("jimm.checkModelMoved")
```
Op

ale8k commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,145 @@
+// one controller managed by JIMM to another controller managed by JIMM.
+func (j *JujuManager) checkModelMigratedInternal(ctx context.Context, api API, m *dbmodel.Model) {
+	const op = errors.Op("jimm.checkModelMoved")
+	zapctx.Info(ctx, string(op))
```
zapctx.WithFields be better to always set model uuid

ale8k commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,145 @@
+		// failed to migrate otherwise we'd expect a redirect error.
+		if modelInfo.Migration.End != nil {
+			m.ProcessFailedMigration()
+			err = j.Database.UpdateModel(ctx, m)
```
Can we inline these error calls, it'll reduce LoC and for me personally it is easier to read

kian99 replied:
Sure

ale8k commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,145 @@
+			m.ProcessFailedMigration()
+			err = j.Database.UpdateModel(ctx, m)
+			if err != nil {
+				zapctx.Error(ctx, "failed to update model after failed migration", zap.String("model", m.UUID.String), zap.Error(err))
```
We can return an error in this function and have a single zapctx.Error in PollModels, there's no reason this couldn't return an error and the caller should define the behaviour of what to do in the event of an error. 

tldr bubble errors up please.

ale8k commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,145 @@
+		return
+	}
+
+	// Expect a redirect error if the model successfully migrated.
```
Lines 87 to 106 would be good in a fun called handleRedirectInfo or something similar with a godoc explaining why we're rummaging through the error to get the redirect info. This isn't super immediately clear why it must be done like this from reading and whilst I understand what you're doing here, it definitely looks nuanced and is worth a godoc.

Then it'd be good to pass that redirectInfo to another func called getNewControllerName with another godoc detailing why we use alias and what it really means to us using the alias. I'd expect this to just return a string.

kian99 replied:
I'd prefer to keep it inline and it can have an inline comment, it's short enough that keeping it all in one place helps.

ale8k reviewed (changes requested):
Some minor changes, but after that lgtm.

kian99 committed:
refactor: bubble up errors in model poller

Also did minor renaming.

ale8k commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,158 @@
+		// - Do nothing if the model is in any other migration mode (MigrationModeImporting, MigrationModeExporting)
+		for _, m := range models {
+			ctx := zapctx.WithFields(ctx,
+				zap.String("modelOwner", m.OwnerIdentityName),
```
We haven't got much consistency on label keys, I tend to use hyphens/snake and this is camel. What do you think about sticking with "model-owner" etc for the labels?

kian99 replied:
Yeah we don't have much consistency, I'll change it to model-owner, but I've seen camel case in some places.

kian99 committed:
feat: add ForUpdate() DB method

kian99 committed:
chore: update camelCase to bbq-case for log keys

alesstimec commented on internal/db/modifiers.go:
```diff
@@ -0,0 +1,24 @@
+//	                err = tx.UpdateModel(ctx, &model)
+//					return nil
+//			})
+func (d *Database) ForUpdate() *Database {
```
this one is a bit scary, because it's so easy to hold it wrong as it does not limit you to a single select, or a select at all..  but then again, we can always argue it's a "skill  issue"

alesstimec commented on internal/dbmodel/model.go:
```diff
@@ -182,3 +182,29 @@ func (m Model) MergeModelSummaryFromController(modelSummaryFromController *jujup
+
+// ProcessFailedMigration processes a failed migration by resetting the
+// MigrationMode to MigrationModeNone.
+func (m *Model) ProcessFailedMigration() {
```
s/ProcessFailedMigration/MigrationFailed/ or SetMigrationFailed

kian99 replied:
Will go with `MigrationFailed`.

alesstimec commented on internal/dbmodel/model.go:
```diff
@@ -182,3 +182,29 @@ func (m Model) MergeModelSummaryFromController(modelSummaryFromController *jujup
+
+// ProcessSuccessfulInternalMigration processes a successful internal migration
+// by setting the MigrationMode to MigrationModeNone and updating the ControllerID.
+func (m *Model) ProcessSuccessfulInternalMigration(controllerID uint) {
```
SetSuccessfulInternalMigration

kian99 replied:
To match the above, I'll go with `InternalMigrationSuccess`.

alesstimec commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	// We expect this controller will be known to JIMM.
+	controller := dbmodel.Controller{Name: redirectInfo.ControllerAlias}
```
please parse the `redirectInfo.ControllerTag` to get the controller UUID and use that.. i have no idea where this alias comes from and i'd be much more comfortable relying on UUID to identify the controller

kian99 replied:
As discussed offline, Juju's model manager facade doesn't return the controller tag,
https://github.com/juju/juju/blob/3.6/apiserver/common/modelmanagerinterface.go#L208
But model level facades do return the tag https://github.com/juju/juju/blob/3.6/apiserver/admin.go#L386
In internal/jimm we don't have a mechanism for calling model endpoints (i.e. /mode/uuid/api) so we'll stick with the controller alias.

alesstimec commented on internal/jimm/juju/model_poller.go:
```diff
@@ -0,0 +1,158 @@
+// This performs eventual cleanup of models that have been deleted through
+// the API (since the deletion of a model is not immediate) and handles
+// cases where the model was deleted directly on the underlying controller.
+func (j *JujuManager) maybeCleanupModel(ctx context.Context, api API, m *dbmodel.Model) error {
```
cleanupDeletedModels?

kian99 replied:
Went with Alex's suggestion for maybCleanupModel which I think is fine.

alesstimec reviewed (approved):
LGTM, with a few comments. please use the controller UUID in the redirect error to identify the target controller

kian99 committed:
chore: rename dbmodel methods

ale8k reviewed (approved):
lgtm now, ty for changes man!
